src/frontend/AppContainerWrapper.tsx

```diff
@@ -1,189 +1,105 @@
+        if (initialUrl) {
+          setInitialNavState(undefined);
+          return;
+        }
 
-      if (previousRouteName !== currentRouteName) {
-        updateRouteBasedAppState(currentRouteName);
+        setInitialNavState(await loadSavedNavState(log));
+      } catch {
+        // handle error here
       }
-    },
-    [updateRouteBasedAppState]
-  );
-
-  const openInviteModal = React.useCallback((key: string) => {
-    if (navRef.current) {
-      navRef.current.dispatch(
-        NavigationActions.navigate({
-          routeName: "ProjectInviteModal",
-          params: { invite: key },
-        })
-      );
     }
   }, []);
 
-  const { loadNavigationState, persistNavigationState } = React.useMemo(
-    () =>
-      IS_E2E
-        ? {}
-        : {
-            loadNavigationState: async () => {
-              const loadedNavState = await loadSavedNavState();
-
-              const loadedRouteName = getRouteName(loadedNavState);
-
-              updateRouteBasedAppState(loadedRouteName);
-
-              return loadedNavState;
-            },
-            persistNavigationState: createNavigationStatePersister(),
-          },
-    [updateRouteBasedAppState]
+  const handleNavStateChange = React.useCallback(
+    (navState?: NavigationState) => {
+      if (!navState || IS_E2E) return;
+      setCurrentRoute(getRouteName(navState));
+      persistNavigationState(log, navState);
+    },
+    []
   );
 
-  /**
-   * TODO: Uncomment when project invites are supported
-   */
-  //   useProjectInviteListener(invite => {
-  //     if (inviteModalEnabled) {
-  //       openInviteModal(invite);
-  //     } else {
-  //       setQueuedInvite(invite);
-  //     }
-  //   });
+  const inviteModalEnabled = inviteModalDisabledOnRoute(currentRoute);
+  const hidePracticeBar = hidePracticeBarForRoute(currentRoute);
+  const hidePracticeMode = hidePracticeModeTemporarily(currentRoute);
 
   React.useEffect(() => {
     if (inviteModalEnabled && queuedInvite) {
       setQueuedInvite(null);
-      openInviteModal(queuedInvite);
+      navRef.navigate("ProjectInviteModal", { inviteKey: queuedInvite });
     }
-  }, [inviteModalEnabled, queuedInvite, openInviteModal]);
+  }, [navRef, inviteModalEnabled, queuedInvite]);
+
+  if (initialNavState === "loading") {
+    return <Loading />;
+  }
 
   return (
     <PracticeMode
       enabled={devExperiments.onboarding && !hidePracticeMode}
       hideBar={hidePracticeBar}
     >
-      <AppContainer
-        loadNavigationState={loadNavigationState}
-        onNavigationStateChange={onNavStateChange}
-        persistNavigationState={persistNavigationState}
-        ref={nav => {
-          if (nav) {
-            navRef.current = nav;
-          }
-        }}
-        uriPrefix={URI_PREFIX}
-      />
+      <NavigationContainer
+        initialState={initialNavState}
+        onStateChange={handleNavStateChange}
+        linking={{ prefixes: [URI_PREFIX] }}
```

Comment:
[Don't we need](https://reactnavigation.org/docs/navigation-container#ref) `ref={navRef}` here?